Unify error handling of bad dictionary elements

cmb69 opened the pull request:
As is, we have separate error handling and reporting for empty strings and strings containing NUL bytes.  For simpler error handling we merge both cases; while that is slightly worse regarding the error reporting, it shouldn't matter in practice, since users need to check the validity of the dictionary upfront anyway.

---

This is a follow-up to PR #16335, where @TimWolla and @Girgias made suggestions for further improvements. This is the "best" I could come up with.

Girgias commented on ext/zlib/zlib.c:
```diff
@@ -808,24 +808,15 @@ static bool zlib_create_dictionary_string(HashTable *options, char **dict, size_
-						ZEND_ASSERT(*ptr);
-						if (ZSTR_LEN(*ptr) == 0 || EG(exception)) {
+						*++ptr = zval_try_get_string(cur);
+						if (*ptr == NULL || ZSTR_LEN(*ptr) == 0 || zend_str_has_nul_byte(*ptr)) {
```
Something else I'm noticing, is that if the `option_buffer` zval is a string we don't check if it doesn't have a nul byte AFAICS.

cmb69 replied:
That is correct, because a dictionary string is a char array of zero terminated strings (otherwise you could only pass a single word).

Girgias reviewed (commented):
I don't mind, but I also don't understand why you don't use a goto to some clean-up section?

cmb69 commented:
> I also don't understand why you don't use a goto to some clean-up section?

Something like:
````diff
 ext/zlib/zlib.c | 30 ++++++++++++++----------------
 1 file changed, 14 insertions(+), 16 deletions(-)

diff --git a/ext/zlib/zlib.c b/ext/zlib/zlib.c
index 4221bceacc..4ece73ab46 100644
--- a/ext/zlib/zlib.c
+++ b/ext/zlib/zlib.c
@@ -808,25 +808,14 @@ static bool zlib_create_dictionary_string(HashTable *options, char **dict, size_
 					zend_string **end, **ptr = strings - 1;
 
 					ZEND_HASH_FOREACH_VAL(dictionary, cur) {
-						*++ptr = zval_get_string(cur);
-						ZEND_ASSERT(*ptr);
-						if (ZSTR_LEN(*ptr) == 0 || EG(exception)) {
-							do {
-								zend_string_release(*ptr);
-							} while (--ptr >= strings);
-							efree(strings);
-							if (!EG(exception)) {
-								zend_argument_value_error(2, "must not contain empty strings");
-							}
-							return 0;
+						*++ptr = zval_try_get_string(cur);
+						if (*ptr == NULL || ZSTR_LEN(*ptr) == 0) {
+							zend_argument_value_error(2, "must not contain empty strings");
+							goto fail;
 						}
 						if (zend_str_has_nul_byte(*ptr)) {
-							do {
-								zend_string_release(*ptr);
-							} while (--ptr >= strings);
-							efree(strings);
 							zend_argument_value_error(2, "must not contain strings with null bytes");
-							return 0;
+							goto fail;
 						}
 
 						*dictlen += ZSTR_LEN(*ptr) + 1;
@@ -842,6 +831,15 @@ static bool zlib_create_dictionary_string(HashTable *options, char **dict, size_
 						zend_string_release_ex(*ptr, 0);
 					} while (++ptr != end);
 					efree(strings);
+					break;
+fail:
+					do {
+						if (*ptr != NULL) {
+							zend_string_release(*ptr);
+						}
+					} while (--ptr >= strings);
+					efree(strings);
+					return 0;
 				}
 			} break;
 
````

In my opinion, that's harder to understand, and error-prone (if you forget the `break`, you have double-frees). And actually we need a third error for the case where the array element cannot be converted to string; that's yet another leg.

Girgias commented:
> > I also don't understand why you don't use a goto to some clean-up section?
> 
> Something like:
> 
> ```diff
>  ext/zlib/zlib.c | 30 ++++++++++++++----------------
>  1 file changed, 14 insertions(+), 16 deletions(-)
> 
> diff --git a/ext/zlib/zlib.c b/ext/zlib/zlib.c
> index 4221bceacc..4ece73ab46 100644
> --- a/ext/zlib/zlib.c
> +++ b/ext/zlib/zlib.c
> @@ -808,25 +808,14 @@ static bool zlib_create_dictionary_string(HashTable *options, char **dict, size_
>  					zend_string **end, **ptr = strings - 1;
>  
>  					ZEND_HASH_FOREACH_VAL(dictionary, cur) {
> -						*++ptr = zval_get_string(cur);
> -						ZEND_ASSERT(*ptr);
> -						if (ZSTR_LEN(*ptr) == 0 || EG(exception)) {
> -							do {
> -								zend_string_release(*ptr);
> -							} while (--ptr >= strings);
> -							efree(strings);
> -							if (!EG(exception)) {
> -								zend_argument_value_error(2, "must not contain empty strings");
> -							}
> -							return 0;
> +						*++ptr = zval_try_get_string(cur);
> +						if (*ptr == NULL || ZSTR_LEN(*ptr) == 0) {
> +							zend_argument_value_error(2, "must not contain empty strings");
> +							goto fail;
>  						}
>  						if (zend_str_has_nul_byte(*ptr)) {
> -							do {
> -								zend_string_release(*ptr);
> -							} while (--ptr >= strings);
> -							efree(strings);
>  							zend_argument_value_error(2, "must not contain strings with null bytes");
> -							return 0;
> +							goto fail;
>  						}
>  
>  						*dictlen += ZSTR_LEN(*ptr) + 1;
> @@ -842,6 +831,15 @@ static bool zlib_create_dictionary_string(HashTable *options, char **dict, size_
>  						zend_string_release_ex(*ptr, 0);
>  					} while (++ptr != end);
>  					efree(strings);
> +					break;
> +fail:
> +					do {
> +						if (*ptr != NULL) {
> +							zend_string_release(*ptr);
> +						}
> +					} while (--ptr >= strings);
> +					efree(strings);
> +					return 0;
>  				}
>  			} break;
>  
> ```
> 
> In my opinion, that's harder to understand, and error-prone (if you forget the `break`, you have double-frees). And actually we need a third error for the case where the array element cannot be converted to string; that's yet another leg.

Basically, I need to pray for C2Y to give us `defer` :D but I don't mind either way.

cmb69 commented:
> Basically, I need to pray for C2Y to give us `defer` :D but I don't mind either way.

Would that work block scoped? Because I guess that is my actual issue with a goto here.

cmb69 commented:
Thank you @TimWolla! This is really nice! I have replaced my suggestion with yours, and added small commit on top, which @Girgias will like (I think).

The only problem is that there is still no `goto` … ;)

TimWolla commented:
> This is really nice! I have replaced my suggestion with yours,

Thanks. I only have one small request: Please do not @-mention me in commits. This annoyingly sends a notification email whenever the commit is pushed anywhere (including rebases and cherry-picks).

My preferred way of getting credit is adding this at the bottom of the commit message:

```
Co-authored-by: Tim Düsterhus <tim@bastelstu.be>
```

cmb69 commented:
> My preferred way of getting credit is adding this at the bottom of the commit message:
> 
> ```
> Co-authored-by: Tim Düsterhus <tim@bastelstu.be>
> ```

Okay, will change that. However, it is actually the other way round. :)

TimWolla commented on ext/zlib/zlib.c:
```diff
@@ -796,61 +796,64 @@ static bool zlib_create_dictionary_string(HashTable *options, char **dict, size_
+							result = 0;
 							zend_argument_value_error(2, "must not contain strings with null bytes");
-							return 0;
+							break;
```
I just realize that this will leak `string`: It's not yet added to the `strings` array. You'll probably need to move the lines 829-831 up to 817.

cmb69 replied:
Indeed. Thank you! Should be fixed now.

Do you know if there is a way to force the engine to not use interned strings at all? That could be useful to detect such leaks.

TimWolla replied:
> Do you know if there is a way to force the engine to not use interned strings at all?

I'm not aware of anything.

TimWolla commented:
One thing that I remembered exists after my initial suggestion: The `smart_string` API (that's the one working on char* instead of zend_string*) might be able to simplify this further. I have this wonderful piece of code, which also comes with the `goto`. I've previously used the `if(0)` trick in https://github.com/php/php-src/blob/8f1543a76e61f8a8d353c36bb31f3f21c389ee53/Zend/zend_vm_def.h#L4214-L4215.

```
			case IS_ARRAY: {
				HashTable *dictionary = Z_ARR_P(option_buffer);
				bool result = 1;

				if (zend_hash_num_elements(dictionary) > 0) {
					zend_string *string;

					zval *cur;
					smart_string out = {0};
					ZEND_HASH_FOREACH_VAL(dictionary, cur) {
						string = zval_try_get_string(cur);
						if (string == NULL) {
							string = ZSTR_EMPTY_ALLOC();
							goto fail;
						}

						smart_string_appendl(&out, ZSTR_VAL(string), ZSTR_LEN(string));
						smart_string_appendc(&out, '\0');

						if (ZSTR_LEN(string) == 0) {
							zend_argument_value_error(2, "must not contain empty strings");
							goto fail;
						}
						if (zend_str_has_nul_byte(string)) {
							zend_argument_value_error(2, "must not contain strings with null bytes");
							goto fail;
						}

						zend_string_release(string);
					} ZEND_HASH_FOREACH_END();

					*dict = out.c;
					*dictlen = out.len;

					if (0) {
 fail:
						result = 0;
						zend_string_release(string);
						smart_string_free(&out);
					}
				}

				return result;
			}
```

cmb69 commented:
> The `smart_string` API (that's the one working on char* instead of zend_string*) might be able to simplify this further.

Might be a good idea, although for a long time I'm having an issue with the way the underlying buffer is resized; that doesn't look smart to me.

> I have this wonderful piece of code, which also comes with the `goto`.

Fortunately, my therapist helped: "breathe – imagine a green field – breathe – imagine warm sunshine – breathe". That went on for hours, but finally I'm well again.
